pkg/cmd/pulumi/up.go

```diff
@@ -111,23 +111,28 @@ func newUpCmd() *cobra.Command {
 			return result.FromError(fmt.Errorf("getting stack configuration: %w", err))
 		}
 
-		targetURNs := []resource.URN{}
-		snap, err := s.Snapshot(ctx)
-		if err != nil {
-			return result.FromError(err)
-		}
-		for _, t := range targets {
-			targetURNs = append(targetURNs, snap.GlobUrn(resource.URN(t))...)
-		}
+		targetURNs, replaceURNs := []resource.URN{}, []resource.URN{}
 
-		replaceURNs := []resource.URN{}
-		for _, r := range replaces {
-			replaceURNs = append(replaceURNs, snap.GlobUrn(resource.URN(r))...)
-		}
+		if len(targets)+len(replaces)+len(targetReplaces) > 0 {
+			// This call to s.Snapshot adds latency s.Update below will call s.Snapshot again, presumably
```

Comment:
It shouldn't cloudStack.Snapshot is a caching call it will only make one http request.

Reply:
That is, if your seeing two calls to exportStack its because the stack object "s" isn't being shared.

Reply:
I'm seeing 2 http requests in the trace. Tracing stacks of the cloud backend ExportStackDeployment seeing this:

```

pulumi up --yes --skip-preview 

- *first*

runtime/debug.Stack()
        /nix/store/72g8s7d0izsx4j1vhfssds3vbpkby7nv-go-1.18.2/share/go/src/runtime/debug/stack.go:24 +0x65
github.com/pulumi/pulumi/pkg/v3/backend/httpstate/client.(*Client).ExportStackDeployment(0xc000bcfa40, {0x5e8b3a8, 0xc00011e000}, {{0xc000a3e450, 0x5}, {0xc00141c0b0, 0x5}, {0xc000a3e45a, 0x3}}, 0x0)
        /Users/anton/pulumi/pkg/backend/httpstate/client/client.go:475 +0x73
github.com/pulumi/pulumi/pkg/v3/backend/httpstate.(*cloudBackend).exportDeployment(0xc000dbe5a0, {0x5e8b3a8, 0xc00011e000}, {0x5e86870?, 0xc000df30a8?}, 0xc000494001?)
        /Users/anton/pulumi/pkg/backend/httpstate/backend.go:1334 +0x12b
github.com/pulumi/pulumi/pkg/v3/backend/httpstate.(*cloudBackend).getSnapshot(0xc000df30a0?, {0x5e8b3a8, 0xc00011e000}, {0x5e86870?, 0xc000df30a8?})
        /Users/anton/pulumi/pkg/backend/httpstate/state.go:218 +0x30
github.com/pulumi/pulumi/pkg/v3/backend/httpstate.(*cloudStack).Snapshot(0xc000aa0310, {0x5e8b3a8, 0xc00011e000})
        /Users/anton/pulumi/pkg/backend/httpstate/stack.go:126 +0xab
main.newUpCmd.func1({_, _}, {{{0x0, 0x0, 0x0}, {0x0, 0x0, 0x0}, 0x0, 0x0, ...}, ...})
        /Users/anton/pulumi/pkg/cmd/pulumi/up.go: **115** +0x630
main.newUpCmd.func3(0x0?, {0xc0009e9960, 0x0, 0x0?})
        /Users/anton/pulumi/pkg/cmd/pulumi/up.go:458 +0x665
github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil.RunResultFunc.func1(0xc000c75900?, {0xc0009e9960, 0x0, 0x2})
        /Users/anton/pulumi/sdk/go/common/util/cmdutil/exit.go:112 +0x4f
github.com/spf13/cobra.(*Command).execute(0xc000c75900, {0xc0009e9940, 0x2, 0x2})
        /Users/anton/go/pkg/mod/github.com/spf13/cobra@v1.5.0/command.go:876 +0x67b
github.com/spf13/cobra.(*Command).ExecuteC(0xc000ba3b80)
        /Users/anton/go/pkg/mod/github.com/spf13/cobra@v1.5.0/command.go:990 +0x3b4
github.com/spf13/cobra.(*Command).Execute(...)
        /Users/anton/go/pkg/mod/github.com/spf13/cobra@v1.5.0/command.go:918
main.main()
        /Users/anton/pulumi/pkg/cmd/pulumi/main.go:48 +0x45

-*second*

Calling ExportStackDeployment with this stack: goroutine 1 [running]:
runtime/debug.Stack()
        /nix/store/72g8s7d0izsx4j1vhfssds3vbpkby7nv-go-1.18.2/share/go/src/runtime/debug/stack.go:24 +0x65
github.com/pulumi/pulumi/pkg/v3/backend/httpstate/client.(*Client).ExportStackDeployment(0xc000bcfa40, {0x5e8b3a8, 0xc00011e000}, {{0xc000a3e450, 0x5}, {0xc001772140, 0x5}, {0xc000a3e45a, 0x3}}, 0x0)
        /Users/anton/pulumi/pkg/backend/httpstate/client/client.go:475 +0x73
github.com/pulumi/pulumi/pkg/v3/backend/httpstate.(*cloudBackend).exportDeployment(0xc000dbe5a0, {0x5e8b3a8, 0xc00011e000}, {0x5e86870?, 0xc0007920c0?}, 0xc000dbe5a0?)
        /Users/anton/pulumi/pkg/backend/httpstate/backend.go:1334 +0x12b
github.com/pulumi/pulumi/pkg/v3/backend/httpstate.(*cloudBackend).getSnapshot(0xec403c0?, {0x5e8b3a8, 0xc00011e000}, {0x5e86870?, 0xc0007920c0?})
        /Users/anton/pulumi/pkg/backend/httpstate/state.go:218 +0x30
github.com/pulumi/pulumi/pkg/v3/backend/httpstate.(*cloudBackend).getTarget(0x5e8b3a8?, {0x5e8b3a8, 0xc00011e000}, {0x5e86870, 0xc0007920c0}, 0xc000750600, {0xee19730?, 0x6d10548})
        /Users/anton/pulumi/pkg/backend/httpstate/state.go:238 +0x105
github.com/pulumi/pulumi/pkg/v3/backend/httpstate.(*cloudBackend).newUpdate(_, {_, _}, {_, _}, {0xc00047f2c0, {0xc000058240, 0x19}, {0x0, 0x0, ...}, ...}, ...)
        /Users/anton/pulumi/pkg/backend/httpstate/state.go:200 +0x1d8
github.com/pulumi/pulumi/pkg/v3/backend/httpstate.(*cloudBackend).runEngineAction(_, {_, _}, {_, _}, {_, _}, {0xc00047f2c0, {0xc000058240, 0x19}, ...}, ...)
        /Users/anton/pulumi/pkg/backend/httpstate/backend.go:1082 +0x133
github.com/pulumi/pulumi/pkg/v3/backend/httpstate.(*cloudBackend).apply(_, {_, _}, {_, _}, {_, _}, {0xc00047f2c0, {0xc000058240, 0x19}, ...}, ...)
        /Users/anton/pulumi/pkg/backend/httpstate/backend.go:1047 +0x41f
github.com/pulumi/pulumi/pkg/v3/backend.PreviewThenPromptThenExecute({_, _}, {_, _}, {_, _}, {0xc00047f2c0, {0xc000058240, 0x19}, {0x0, ...}, ...}, ...)
        /Users/anton/pulumi/pkg/backend/apply.go:237 +0x24b
github.com/pulumi/pulumi/pkg/v3/backend/httpstate.(*cloudBackend).Update(_, {_, _}, {_, _}, {0xc00047f2c0, {0xc000058240, 0x19}, {0x0, 0x0, ...}, ...})
        /Users/anton/pulumi/pkg/backend/httpstate/backend.go:930 +0xc5
github.com/pulumi/pulumi/pkg/v3/backend.UpdateStack({_, _}, {_, _}, {0xc00047f2c0, {0xc000058240, 0x19}, {0x0, 0x0, 0x0}, ...})
        /Users/anton/pulumi/pkg/backend/stack.go:91 +0xba
github.com/pulumi/pulumi/pkg/v3/backend/httpstate.(*cloudStack). **Update** (_, {_, _}, {0xc00047f2c0, {0xc000058240, 0x19}, {0x0, 0x0, 0x0}, 0xc0002c83c0, ...})
        /Users/anton/pulumi/pkg/backend/httpstate/stack.go:152 +0x88
main.newUpCmd.func1({_, _}, {{{0x6d10548, 0x0, 0x0}, {0x0, 0x0, 0x0}, 0x7fffffff, 0x0, ...}, ...})
        /Users/anton/pulumi/pkg/cmd/pulumi/up.go: **180** +0xd37
main.newUpCmd.func3(0x0?, {0xc0009e9960, 0x0, 0x0?})
        /Users/anton/pulumi/pkg/cmd/pulumi/up.go:458 +0x665
github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil.RunResultFunc.func1(0xc000c75900?, {0xc0009e9960, 0x0, 0x2})
        /Users/anton/pulumi/sdk/go/common/util/cmdutil/exit.go:112 +0x4f
github.com/spf13/cobra.(*Command).execute(0xc000c75900, {0xc0009e9940, 0x2, 0x2})
        /Users/anton/go/pkg/mod/github.com/spf13/cobra@v1.5.0/command.go:876 +0x67b
github.com/spf13/cobra.(*Command).ExecuteC(0xc000ba3b80)
        /Users/anton/go/pkg/mod/github.com/spf13/cobra@v1.5.0/command.go:990 +0x3b4
github.com/spf13/cobra.(*Command).Execute(...)
        /Users/anton/go/pkg/mod/github.com/spf13/cobra@v1.5.0/command.go:918
main.main()
        /Users/anton/pulumi/pkg/cmd/pulumi/main.go:48 +0x45

```

Reply:
This line perhaps:

https://github.com/pulumi/pulumi/blob/e96c4ef162fffcbe4c501b9fe3900010c343da8f/pkg/backend/httpstate/backend.go#L1047

Stack becomes StackRef so the cache drops. It then calls `(*cloudBackend).getSnapshot` through a different pathway.

Reply:
I'm inclined to let that be and check in this change as is as a quick win (though I can amend the doc comment with additional detail). I can add a test to lock in the fact we call the function only once during a typical update if we later refactor this to share Stack object and its cache. WDYT?

Reply:
Yeh the change seems fine, just amend the comment to make it clear that its the underlying http call being duplicated and not the stack.Snapshot call.